### PR TITLE
[lldb][llvm] Return an error instead of crashing when parsing a line table prologue.

### PR DESCRIPTION
**Pull request by clayborg**
We recently ran into some bad DWARF where the `DW_AT_stmt_list` of many compile units was randomly set to invalid values and was causing LLDB to crash due to an assertion about address sizes not matching. Instead of asserting, we should return an appropriate recoverable `llvm::Error`.

### CODE REVIEW

File: llvm/lib/DebugInfo/DWARF/DWARFDebugLine.cpp [hunk 1]
```diff
@@ -372,26 +372,56 @@ Error DWARFDebugLine::Prologue::parse(
...
+      }
+    } else if (DataAddrSize != getAddressSize()) {
+      return ReportErrorOrCursor(createStringError(
+          errc::not_supported,
+          "parsing line table prologue at offset 0x%8.8" PRIx64 ": address "
+          "size %" PRIu8 " doesn't match architecture address size %" PRIu8,
+          PrologueOffset, PrologueAddrSize, DataAddrSize));
+    }
     SegSelectorSize = DebugLineData.getU8(Cursor);
   }
```
Comment (dwblaikie) [inline llvm/lib/DebugInfo/DWARF/DWARFDebugLine.cpp:426]: These should be recoverable errors, because we can skip this line table and read the next one.

**Comment by clayborg** [comment]
I have made the prologue errors recoverable. Anything else needed here?

**Review by JDevlieghere — APPROVED** [review]
LGTM!

File: llvm/lib/DebugInfo/DWARF/DWARFDebugLine.cpp [hunk 2]
```diff
@@ -389,9 +389,25 @@ Error DWARFDebugLine::Prologue::parse(
...
+      if (DataAddrSize == 0) {
+        if (PrologueAddrSize != 4 && PrologueAddrSize != 8) {
+          RecoverableErrorHandler(createStringError(
+              errc::not_supported,
+              "parsing line table prologue at offset 0x%8.8" PRIx64
+              ": invalid address size %" PRIu8,
+              PrologueOffset, PrologueAddrSize));
+        }
+      } else if (DataAddrSize != PrologueAddrSize) {
+        RecoverableErrorHandler(createStringError(
```
Comment (dwblaikie) [inline llvm/lib/DebugInfo/DWARF/DWARFDebugLine.cpp:404]: Looks like this path isn't tested? Could you add a test?

File: llvm/unittests/DebugInfo/DWARF/DWARFDebugLineTest.cpp [hunk 3]
```diff
@@ -823,7 +823,9 @@ TEST_F(DebugLineBasicFixture, ErrorForUnsupportedAddressSizeDefinedInHeader) {
                                                     nullptr, RecordRecoverable);
   EXPECT_THAT_ERROR(
       std::move(Recoverable),
-      FailedWithMessage("address size 0x09 of DW_LNE_set_address opcode at "
```
Comment (dwblaikie) [inline llvm/unittests/DebugInfo/DWARF/DWARFDebugLineTest.cpp:826]: Looks like this test change removed coverage for a different error message, by failing earlier?

This test should probably be restored by modifying it to be reachable, or the original code's error handling should be turned into an assertion if the diagnostic is now an invariant for the code?

**Pull request merged**